README.md
# README

This repository contains the scripts used to: 1) mine taxonomy data from the NCBI taxonomy database and from the International Union for Conservation of Nature website, 2) mine COI sequence records from the NCBI nucleotide database, BOLD API, and BOLD data releases.  I also provide additional scripts to filter results, print reports, generate figures, etc.

## How to cite

If you find these methods/scripts useful and use them in a publication, please cite:
Porter, T.M., Hajibabaei, M. (2018) Over 2.5 million COI sequences and growing. bioRxiv, DOI: 10.1101/353904

## Overview

[Part I - Retrieve Taxonomy IDs from the NCBI taxonomy database](#part-i---retrieve-taxonomy-ids-from-the-ncbi-taxonomy-database)  
[Part II - Retrieve COI records from the NCBI nucleotide database](#part-ii---retrieve-coi-records-from-the-ncbi-nucleotide-database)  
[Part III - Parse through the COI GenBank records and print reports](#part-iii---parse-through-the-coi-genbank-records-and-print-reports)  
[Part IV - IUCN endangered animal species dataset](#part-iv---iucn-endangered-animal-species-dataset)  
[Part V - BARCODE dataset](#part-v---barcode-dataset)  
[Part VI - Freshwater dataset](#part-vi---freshwater-dataset)  
[Part VII - Map country & latlon metadata with R](#part-vii---map-country---latlon-metadata-with-R)  
[Part VIII - Other R figures](#part-viii---other-r-figures)  
[Part IX - Retrieve BOLD data via the BOLD API](#part-ix---retrieve-bold-data-via-the-bold-api)  
[Part X - Retrieve BOLD data via the data releases](#part-x---retrieve-bold-data-via-the-data-releases)

## Part I - Retrieve Taxonomy IDs from the NCBI taxonomy database

Retrieve all taxonids from the NCBI taxonomy database for all Eukaryotes with a species name.  Be sure to add your email address to line 85 in ebot_taxonomy3.plx.

```perl
perl ebot_taxonomy3.plx
```

The outfile is taxonomy.taxid which contains a list of taxonids.

For each taxonid, skip entries that contain sp., nr., aff., or cf. else retrieve the genus and species.  Be sure to update the hard-coded paths to names.dmp and nodes.dmp on lines 29 and 30 in taxonomy_crawl_for_genus_species_list.plx  These can be obtained from ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz and should be kept current.

```linux
perl taxonomy_crawl_for_genus_species_list.plx taxonomy.taxid
```

The outfile is Genus_species.txt

Use the linux split command to break Genus_species.txt into smaller files for easier parsing.

```linux
split -l 100 Genus_species.txt
```

The outfiles will be automatically named with the prefix 'x'.

Reformat these files for the NCBI Entrez query using Linux commands and GNU parallel.  GNU parallel can be obtained from https://www.gnu.org/software/parallel/ .  Be sure to update the -j flag to reflect the number of available cores for this step.

```linux
ls | grep '^x' | parallel -j 23 "perl reformat_list_for_entrez_taxonomy.plx {}"
```

Outfiles will be automatically named with the prefix 'x' and file extension .txt

The following commands will clean up the directory a bit

```linux
mkdir reformatted_taxids
mv *.txt reformatted_taxids/.
mv reformatted_taxids/Genus_species.txt .
mkdir original_split_taxids
mv x* original_split_taxids/.
cd reformatted_taxids
```

## Part II - Retrieve COI records from the NCBI nucleotide database

Retrieve GenBank formatted text files from the NCBI nucleotide database.  Use Linux and GNU parallel to run a single job with a bunch of infiles.  The infiles here are the reformatted xsomething.txt files.  This script requires Bio::DB::Eutilities that can be obtained from CPAN.  Be sure to update the search terms used for the Entrez query on line 28 and your email address on line 31. 

```linux
ls | grep .txt | parallel -j 1 "perl grab_many_gb_catch_errors_auto_CO1_year.plx {}"
```

The outfiles will be automatically named 'xsomething_seqs.gb'.  Repeat this search as needed using the appropriate search terms varying the year[PDAT] as needed.  Organize outfiles into their own directories ex. genbank_records_year 

## Part III - Parse through the COI GenBank records and print reports

The script below will search the text-formatted fully identified GenBank COI records for records with sequences are at least 500bp in length, with country and/or latlon metadata.  The taxonomy.taxid file is from Part I.  The dirlist.txt should contain a list of directory names ex. genbank_records_year/ with one entry per line for each year.

```linux
perl quick_filter_gb_country_latlon_CO1.plx taxonomy.taxid dirlist.txt
```

The output is country_latlon_year.txt and is a tab delimited file with the following headers: species, GenBank accession, country, latlon, as well as binary fields for 500bp (0=no, 1=yes), country annotation present, latlon annotation present.

The script below will count up the number of fully identified records that are at least 500bp+ and contain country and/or latlon metadata.

```linux
sort -u country_latlon_year.txt | awk 'BEGIN {FS="\t"} {print $6"\t"$7"\t"$8}' | awk '{sum1 +=$1} {sum2 +=$2} {sum3 +=$3} END {print sum1"\t"sum2"\t"sum3}'
```

If needed, concatenate several outfiles using linux and GNU parallel for further processing of tabular data with Python.

```linux
 ls | grep country_latlon_20 | parallel -j 1 "cat {} >> cat.txt"
 ```

Clean up the tabular output to map this data in R.  The infile is cat.txt from above.

```linux
python3 genbank_map.py
```

The outfile is AllEukaryota_gg_latlon.csv 

## Part IV - IUCN endangered animal species dataset
### Grab a list of target species from the IUCN website

Go to http://www.iucnredlist.org . Use the filters for taxonomy-> Animalia, save the search and export as a .csv file.  From this data grab a list of all binomial species names including synonyms.  Exclude names with affinis or sp.nov.  Name this file IUCN_genus_species.txt.

Go to https://www.ncbi.nlm.nih.gov/Taxonomy/TaxIdentifier/tax_identifier.cgi and submit the list IUCN_genus_species.txt

Grab just the unique taxids. Name this file taxonomy.taxid 

Adjust the line endings if necessary. Remove the quotes. Remove spaces.

### Plug these results into the data flow described in Parts I-III above

Using the clean taxonomy.taxid file with the taxonomy_crawl_for_genus_species_list.plx grab a list of good species names as described in Part I above.  Continue with the steps outlined in Part II to retrieve the COI GenBank records for the endangered animal species dataset.

The script quick_filter_gb_country_latlon_CO1.plx is described above in Part III and will search the text-formatted fully identified GenBank COI records for records with COI sequences are at least 500bp in length, with country and/or latlon metadata.  Continue with the remaining steps described in Part III.

Clean up the tabular output to map this data in R.  The infile is cat.txt.

```linux
python3 IUCN_map.py
```

The outfile is IUCN_gg_latlon.csv 

### Grab organism names

The script below is similar to quick_filter_gb_country_latlon_CO1.plx and searches through the text-formatted fully identified GenBank COI records as above but also includes a field in the outfile for the organism name.  Another difference is that there is no filter for target taxa (ex. endangered species) and all taxa are reported.

```linux
perl quick_filter_gb_country_latlon_CO1_organism.plx dirlist.txt
```

The output country_latlon_year.txt is a tab delimited file as described above with an additional field for organism name.

Concatenate several outfiles using linux and GNU parallel, get a list of the unique species.

```linux
ls | grep organism_20 | parallel -j 1 "cat {} >> organism_cat.txt"
awk 'BEGIN {FS="\t"} {print $4}' organism_cat.txt | sort -u >> organism_cat.unique
```

## Part V - BARCODE dataset

Use the grab_many_gb_catch_errors_auto_CO1_year.plx script from Part II above the the appropriate search terms including the "BARCODE"[KYWD] individually for one year at a time or for a range of years as described in the script. 

Parse through the text formatted COI GenBank records using the quick_filter_gb_country_latlon_CO1.plx script described above in Part III with the newly generated taxonomy.taxid and dirlist.txt files.

Count up the number of fully identified records in the country_latlon_year.txt files that are at least 500bp and contain country and/or latlon metadata as described above in Part III. Do this for each year.

Concatenate the country_latlon_year.txt reports into a single file as described above in Part III.

Reformat the data with python for analysis in R.  The infile is cat.txt.

```linux
python3 barcode_map.py
```

The output file is EukaryotaBarcode_gg_latlon.csv

## Part VI - Freshwater dataset

Get a list of all species names for each higher level taxon of interest for freshwater biomonitoring from the NCBI taxonomy database.  Insert the following search term, one at a time, into the ebot_taxonomy3.plx script: Hirudinea, Oligochaeta AND Metazoa, Gastropoda, Coleoptera, Diptera, Ephemeroptera, Megaloptera, Odonata, Plecoptera AND Polyneoptera, Trichoptera, Amphipoda, Isopoda, Polychaeta, Turbellaria.  Rename each taxonomy.taxid outfile so they don't get overwritten, ex. part1_taxonomy.taxid

```linux
perl ebot_taxonomy3.plx
```

Concatenate all the part_taxonomy.taxid files 

```linux
ls | grep part | parallel -j 1 "cat {} >> taxonomy.taxid"
```

Grab the genus and species names for each taxid using the taxonomy_crawl_for_genus_species_lsit.plx script described above in Part I.  The outfile is Genus_species.txt .  Continue with the steps described in Part I above.

Grab the GenBank formatted text files from the NCBI nucleotide database using the grab_many_gb_catch_errors_auto_CO1_year.plx script described above in Part II.  Do this for each year.

Count up the number of fully identified records in the country_latlon_year.txt files that are at least 500bp and contain country and/or latlon metadata as described in Part III.

Concatenate the country_latlon_year.txt reports into a single file as described above in Part III.

Reformat the data with python for analysis in R.  The infile is cat.txt

```linux
python3 freshwater_map.py
```

The output file is freshwater_gg_latlon.csv

## Part VII - Map country & latlon metadata with R

Script to create Fig 4 is F4_maps_061218.R and uses the dataset AllEukaryota_gg_latlon.csv

Script to create Supplementary Fig 3 (S3) is SF3_maps.R and uses the datasets EukaryotaBarcode_gg_latlon.csv, freshwater_gg_latlon.csv, and IUCN_gg_latlon.csv

## Part VIII - Other R figures

Script to create Fig 1 is F1_bars.R and the dataset is F1.csv

Script to create Fig 2 is F2_facetbars.R and the dataset is F2.csv

Script to create Fig 3 is F3_stackedbar.R and the dataset is F3.csv

Script to create Supplementary Fig 1 (S1) is SF1_horizbar.R and the dataset is FS1.csv

Script to create Supplementary Fig 2 (S2) is SF2_bars.R and the dataset is FS2.csv

## Part IX - Retrieve BOLD data via the BOLD API

This script retrieves FASTA files from the BOLD API.

```linux
zsh query_BOLD.sh taxonlist_all.txt
```

Concatenate the FASTA files into a single FASTA file.

```linux
ls | grep fasta | parallel -j 1 "cat {} >> cat_BOLD_API.fasta"
```

## Part X - Retrieve BOLD data via the data releases

This script retrieves BOLD data releases.

```linux
zsh getBOLDdataReleases.sh releases.txt
```

Unzip all the files using Linux commands.

```linux
ls | grep zip | parallel -j 10 "unzip {}"
```

```linux
perl parse_BOLD_data_releases.plx
```

The outfile is cat_BOLD_datareleases.fasta

# Acknowledgements

I would like to acknowedge funding from the Canadian government through the Genomcis Research and Development Initiative (GRDI) EcoBiomics project.

Last updated: June 26, 2018


Datasets/releases.txt
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_6.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_6.25_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_6.00_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_5.75_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_5.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_5.25_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_5.00_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_4.75_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_4.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_4.25_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_4.00_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_3.75_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_3.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_3.25_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_3.0_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_2.75_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_2.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_2.25_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_2.0_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_1.75_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_1.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_1.25_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_1.00_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_0.75_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/NewPackages/iBOL_phase_0.50_COI.tsv.zip
http://www.boldsystems.org/data/datarelease/CanadianBarcodeNet_ver1.txt.zip


Scripts/F4_maps_061218.R
library(tidyverse)
library(gridExtra)
library(grid)
library(scales)

#June 26, 2018 by Teresita M. Porter

#import records from 2003 to 2017 some of which contain lat/lon and country annotations
#df contains spotty lat/lon and country annot
df<-read.csv('AllEukaryota_gg_latlon.csv', header=T)

#add column labels
names(df)<-c("gb","latitude", "longitude","country_alone")

#count number of records per country
#df2 contains one row per country with num records
df2<-as.data.frame(table(df$country_alone))

#add column labels
names(df2)<-c("country_alone","records")

#get lat and lon to draw world country polygons
wm<-map_data("world")
#each group number corresponds to a country

#mege map data with df2 that contains records/country
wm_records <- left_join(wm, df2, by=c("region"="country_alone"))

#set NA to zero
wm_records$records[is.na(wm_records$records)]<-0

#just look at region & records
wm_region_records<-wm_records[,c("region","records")]

#just look at unique region+record
wm_unique<-unique(wm_region_records)

#get quantiles
quantile(wm_unique$records, c(0.05, 0.25, 0.75, 0.95), na.rm=TRUE)
#5% percentile = 0 records, 
#25% percentile = 36 records, 
#75% percentile = 1354 records, 
#95% percentile = 12393 records

#get average
mean(wm_unique$records, na.rm=TRUE)
#average = 3963 records per country

#use quantiles to put country records into bins for discrete color mapping
wm_records$bins <- cut(wm_records$records, 
                        breaks=c(0, 36, 1354, 12393, 400000), 
                        # 1-36, 37-1354, 1355-12393, 12394-400000
                        limits=c(0,400000),
                        labels=c("-25%","Average","+25%","+45%"))
#add a new factor level for countries with no records
levels(wm_records$bins)<-c(levels(wm_records$bins),"No records")
wm_records$bins[is.na(wm_records$bins)]<-"No records"

#########################################################################
#draw map1 with countries colored
mp<-ggplot() +
  theme_void() +
  theme(legend.position = "none") +
  geom_polygon(data=wm_records, aes(x=long, y=lat, group=group, fill=bins), color=NA) +
  coord_equal() +
  scale_fill_manual(values = c('#C44923','#E4A054','#E1DFBD','#5EBEAC','#0C5B61'),
                    limits = c("No records","-25%","Average","+25%", "+45%"),
					          na.value="#C44923") +
  #Tropic of Cancer
  geom_label(aes(x=-180, y=28, label="23º N"), 
           hjust=0, color="black", label.size=NA, size=3) +
  geom_hline(yintercept=23.4369, color="darkgray", linetype="dashed") +
  #Tropic of Capricorn
  geom_label(aes(x=-180, y=-18, label="23º S"), 
             hjust=0, color="black", label.size=NA, size=3) +
  geom_hline(yintercept=-23.4369, color="darkgray", linetype="dashed")
 
# prep for bar chart
#use quantiles to put UNIQUE country records into bins for discrete color mapping
wm_unique$bins <- cut(wm_unique$records, 
                       breaks=c(0, 36, 1354, 12393, 400000), 
                       # 1-36, 37-1354, 1355-12393, 12394-400000
                       limits=c(0,400000),
                       labels=c("-25%","Average","+25%","+45%"))

#add a new factor level for unique countries with no records
levels(wm_unique$bins)<-c(levels(wm_unique$bins),"No records")
wm_unique$bins[is.na(wm_unique$bins)]<-"No records"

#first sort by records (descending)
wm_unique_sorted<-wm_unique[order(wm_unique$records),]

#reindex
wm_unique_sorted$order<-1:nrow(wm_unique_sorted)

######################################################################
# draw bar chart1 with majority of country data, plot everything except top 5%
bar<-ggplot(data=wm_unique_sorted, 
            aes(x=order,
                y=records, 
                fill=bins)) +
  geom_bar(stat="identity") +
  scale_x_continuous(name="Country",limits=c(1,240), labels=NULL) +
  scale_y_continuous(name="Records",
                     breaks=c(0, 5000, 10000, 15000), 
                     labels=c("0","5,000", "10,000", "15,000"),
                     limits=c(0,15000)) +
  theme(plot.margin=margin(0,0,0,2,"cm"),
        axis.ticks.x=element_blank(),
        legend.key.width=unit(0.75,"cm"),
        legend.key.height=unit(0.25,"cm"),
        legend.position=c(0.5,1),
        legend.justification=c(0.5,1),
        legend.margin=margin(t=0,unit='cm'),
        legend.direction="horizontal",
        legend.title=element_text(size=9),
        legend.text=element_text(size=9),
        panel.grid=element_blank(),
        panel.background = element_blank(),
        axis.line.y=element_line(color="black"),
        axis.line.x=element_line(color="black"),
        axis.title.x=element_text(size=9),
        axis.title.y=element_text(size=9)) +
  scale_fill_manual(values = c('#C44923','#E4A054','#E1DFBD','#5EBEAC','#0C5B61'),
                    limits = c("No records","-25%","Average","+25%", "+45%"),
                    labels = NULL,
                    na.value="#C44923") +
  guides(fill = guide_legend(reverse = F, 
                             title="Average records: 3,963\n\U25BC",
                             title.position="top",
                             title.hjust=0.5,
                             label=TRUE,
                             label.hjust=0,
                             label.vjust=1,
                             label.position="bottom")) +
  annotate("text",x=120,y=10500,label="|         |         |         |\n5th   25th   75th   95th\nPercentiles", size=3, hjust=0.5, vjust=0.5)

top5<-tail(wm_unique_sorted, 13)

######################################################################
# draw bar chart2 with just top 5% plotted
bar2<-ggplot(data=top5, 
            aes(x=order,
                y=records, 
                fill=bins)) +
  geom_bar(stat="identity", width=1) +
  scale_x_continuous(name="Country", 
                     breaks=c(240,241,242,243,244,245,246,247,248,249,250,251,252), 
                     labels=c("Italy","Mexico","France","India","Spain","Brazil","Japan","Germany","China","Australia","Costa Rica","USA","Canada")) +
  scale_y_continuous(name="Records", labels=NULL,
                     limits=c(0,500000)) +
  coord_flip() +
  geom_text(aes(label=comma(records)), hjust=-0.15, size=3) +
  theme(plot.margin=margin(0,2,0,0,"cm"),
        axis.ticks.x=element_blank(),
        legend.position="none",
        panel.grid=element_blank(),
        panel.background = element_blank(),
        axis.line.y=element_line(color="black"),
        axis.line.x=element_line(color="black"),
        axis.title.x=element_text(size=9),
        axis.title.y=element_blank()) +
  scale_fill_manual(values = c('#C44923','#E4A054','#E1DFBD','#5EBEAC','#0C5B61'),
                    limits = c("No records","-25%","Average","+25%", "+45%"),
                    na.value="#C44923")

#######################################################################################
#draw map2 with countries outlined but not filled in, plot points where avail only
mp2<-ggplot() +
  geom_polygon(data=wm_records, aes(x=long, y=lat, group=group), color="lightgrey", fill="lightgrey") +
  geom_point(data=df, mapping=aes(x=longitude, y=latitude, shape="."), 
             show.legend=FALSE, color="violetred", size=0.15) +
  theme_void() +
  theme(legend.position = "none",
        legend.title=element_text(size=9),
        legend.key.width=unit(2.5,"cm")) +
  coord_equal() +
  #Tropic of Cancer
  geom_label(aes(x=-180, y=28, label="23º N"), 
             hjust=0, color="black", label.size=NA, size=3) +
  geom_hline(yintercept=23.4369, color="darkgray", linetype="dashed") +
  #Tropic of Capricorn
  geom_label(aes(x=-180, y=-18, label="23º S"), 
             hjust=0, color="black", label.size=NA, size=3) +
  geom_hline(yintercept=-23.4369, color="darkgray", linetype="dashed") 

#File size is ridiculous so print individually and stick them together later after editing

#use cairo_pdf here to render unicode character for filled triangle
cairo_pdf("Fig4.pdf", family="sans")

lay <- rbind(c(1,1),
             c(2,3),
             c(4,4))

grid.arrange(mp, bar, bar2, mp2, layout_matrix=lay)

dev.off()
